version: "3.8"
services:
  users-db:
    container_name: users_db
    restart: always
    image: postgres:16.1-alpine3.18
    volumes:
      - users-db:/var/lib/postgresql/data
    environment:
      - POSTGRES_USER=$DB_USER
      - POSTGRES_PASSWORD=$DB_PASSWORD
      - POSTGRES_DB=$DB_NAME
    healthcheck:
      test: ["CMD-SHELL", "pg_isready -d $DB_NAME -U $DB_USER"]
      interval: 10s
      timeout: 5s
      retries: 5
    networks:
      - internal

  user-monitoring:
    container_name: user_monitoring
    image: prom/prometheus
    ports:
      - $PROM_PORT:9090
    volumes:
      - .:/etc/prometheus
    command: "--config.file=/etc/prometheus/prometheus.yml"
    networks:
      - internal

  user-service-analytics:
    container_name: user_service_analytics
    image: grafana/grafana:latest
    ports:
      - $GRAFANA_PORT:3000
    environment:
      - GF_SECURITY_ADMIN_USER=$GF_ADMIN_USER
      - GF_SECURITY_ADMIN_PASSWORD=$GF_ADMIN_PASSWORD
    depends_on:
      - user-monitoring
    networks:
      - internal

  user-service:
    build: .
    restart: on-failure
    container_name: user_service
    command: sh -c "goose -dir ./sql/schema/ postgres $DB_URL up && go install && user_service"
    volumes:
      - .:/go/src/app
    ports:
      - $PORT:$PORT
    depends_on:
      users-db:
        condition: service_healthy
    env_file: .env
    networks:
      - internal
      - shared-network

volumes:
  users-db:

networks:
  internal:
    driver: bridge
  shared-network:
    external: true
